### PR TITLE
Refine partial_program for new run_program OP

### PR DESCRIPTION
**Pull request by 0x45f**
### PR types

Others
### PR changes

Others
### Describe

Refine partial_program for new run_program OP

### CODE REVIEW

File: paddle/fluid/eager/to_static/run_program_op_node.h [hunk 1]
```diff
@@ -432,6 +436,8 @@ class GradNodeRunProgram : public egr::GradNodeBase {
     fwd_out_names_ = out_names;
   }
 
+  void SetOut(std::vector<paddle::experimental::Tensor *> out) { out_ = out; }
```
Comment (Aurelius84) [inline paddle/fluid/eager/to_static/run_program_op_node.h]:
```suggestion
  void SetOut(const std::vector<paddle::experimental::Tensor *> &out) { out_ = out; }
```

File: python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py [hunk 2]
```diff
@@ -598,12 +607,12 @@ def _create_fake_var():
                          core.VarDesc.VarType.RAW, False)
         ]
     else:
-        return []
         # TODO(jiabin): Support this later
-        # return [
-        #     core.eager.Tensor(core.VarDesc.VarType.FP32, [], "Fake_var",
-        #                 core.VarDesc.VarType.RAW, False)
-        # ]
+        with paddle.fluid.dygraph.guard():
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py]: remove this todo

File: python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py [hunk 3]
```diff
@@ -459,8 +467,9 @@ def _create_scope_vec(self):
             #                                 "program_out_scope",
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py]: remove unused code

File: paddle/fluid/eager/to_static/run_program_op_node.h (hunk 1 above)
Comment (0x45f) [reply paddle/fluid/eager/to_static/run_program_op_node.h]: 已修改感谢

File: python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py (hunk 2 above)
Comment (0x45f) [reply python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py]: _create_fake_var函数内应该不需要加with paddle.fluid.dygraph.guard()。我看了下self.__fake_vars应该只会在PartialProgramLayer.__call__中用到，将self.__fake_vars设置为lazy应该就可以了

**Review by Aurelius84 — APPROVED** [review]
LGTM

**Review by chenwhql — APPROVED** [review]
LGTM for const_cast

**Pull request merged**